File: source/python/examples/filesystems/upload_file_local_to_sftp_absolute_paths.py
.. code-tab:: python

    from xenon import Server, PasswordCredential, CopyMode, Path

    with Server() as xenon:
        # use the sftp file system adaptor to create a file system
        # representation; the remote filesystem requires credentials to log in,
        # so we'll have to create those too.
        credential = PasswordCredential(
                username='xenon',
                password='javagat')

        remote_fs = xenon.create_file_system(
                adaptor='sftp',
                location='localhost:10022',
                credential=credential)

        # use the local file system adaptor to create a file system
        # representation
        local_fs = xenon.create_file_system(
                adaptor='file')

        # define which file to upload
        local_file = Path('/home/tutorial/xenon/sleep.sh')
        remote_path = Path('/home/xenon/')
        remote_file = remote_path.joinpath(local_file.name)

        # create the destination file only if the destination path doesn't
        # exist yet
        mode = CopyMode.CREATE

        # no need to recurse, we're just downloading a file
        recursive = False

        # perform the copy/upload and wait 1000 ms for the successful or
        # otherwise completion of the operation
        copy_id = local_fs.copy(
                local_file, remote_fs, remote_file,
                mode=mode, recursive=recursive)
        copy_status = local_fs.wait_until_done(copy_id, timeout=1000)

        if copy_status.error:
            print(copy_status.error)
        else:
            print('Done')

        # remember to close the FileSystem instances, or use them as
        # context managers
        remote_fs.close()
        local_fs.close()

File: readthedocs/code-tabs/java/src/main/java/nl/esciencecenter/xenon/tutorial/UploadFileLocalToSftpAbsolutePaths.java
package nl.esciencecenter.xenon.tutorial;

import java.util.Map;

import nl.esciencecenter.xenon.XenonException;
import nl.esciencecenter.xenon.credentials.PasswordCredential;
import nl.esciencecenter.xenon.filesystems.CopyMode;
import nl.esciencecenter.xenon.filesystems.CopyStatus;
import nl.esciencecenter.xenon.filesystems.FileSystem;
import nl.esciencecenter.xenon.filesystems.Path;

public class UploadFileLocalToSftpAbsolutePaths {


    public static void main(String[] args) throws XenonException {

        String host = "localhost";
        String port = "10022";
        Map<String, String> properties = null;

        main(host, port, properties);
    }

    public static void main(String host, String port, Map<String, String> properties) throws XenonException {

        // use the local file system adaptor to create a file system representation
        String adaptorLocal = "file";
        FileSystem filesystemLocal = FileSystem.create(adaptorLocal);

        // define what file to upload
        Path fileLocal = new Path("/home/travis/sleep.sh");

        // use the sftp file system adaptor to create a file system representation; the remote
        // filesystem requires credentials to log in, so we'll have to create those too.
        String adaptorRemote = "sftp";
        String location = host + ":" + port;
        String username = "xenon";
        char[] password = "javagat".toCharArray();
        PasswordCredential credential = new PasswordCredential(username, password);
        FileSystem filesystemRemote = FileSystem.create(adaptorRemote, location, credential, properties);

        // define which file to upload to
        Path fileRemote = new Path("/home/xenon/sleep.sh");

        // create the destination file only if the destination path doesn't exist yet
        CopyMode mode = CopyMode.CREATE;

        // no need to recurse, we're just downloading a file
        boolean recursive = false;

        // perform the copy/upload and wait 1000 ms for the successful or otherwise
        // completion of the operation
        String copyId = filesystemLocal.copy(fileLocal, filesystemRemote, fileRemote, mode, recursive);
        long timeoutMilliSecs = 1000;
        CopyStatus copyStatus = filesystemLocal.waitUntilDone(copyId, timeoutMilliSecs);

        // print any exceptions
        if (copyStatus.getException() != null) {
            System.out.println(copyStatus.getException().getMessage());
        } else {
            System.out.println("Done");
        }
    }
}
